core/cache/application.go

```diff
@@ -38,6 +38,8 @@ type Application struct {
 
 // CharmURL returns the charm url string for this application.
 func (a *Application) CharmURL() string {
+	a.mu.Lock()
```

Comment:
So mutexes are a possible solution, but are usually considered an antipattern in Go. (Share by communicating vs communicate by sharing.)

Is there a way we could avoid having multiple goroutines mutating shared state instead?
eg, you get a copy of the cache object, but it isn't expected to state up-to-date. I'm a bit concerned that we'll start getting ourselves into mutex hell and end up with deadlocks, etc. (This goroutine looks at Application and then Unit, and that goroutine looks at Unit and then Application, and you end up deadlocking your mutex operations.)

Again, the actual changes all seem sane in isolation, I'm just checking if we're doing something wrong at a higher level that we should be reconsidering.

I think having Cache.Application() return a *copy* of the Application object, instead of a pointer to the cached value is quite valuable, as it also means arbitrary code can't mutate what you think is in the cache.

Thoughts?